nikola/filters.py

```diff
@@ -391,3 +392,15 @@ def _normalize_html(data):
 
 
 normalize_html = apply_to_text_file(_normalize_html)
+
+
+@apply_to_text_file
+def add_header_permalinks(data):
+    """Post-process HTML via lxml to add header permalinks Sphinx-style."""
+    doc = lxml.html.document_fromstring(data)
+    for h in ['h1', 'h2', 'h3', 'h4']:
+        nodes = doc.findall('*//%s' % h)
+        for node in nodes:
+            new_node = lxml.html.fragment_fromstring('<a id="{0}" href="#{0}" class="headerlink" title="Permalink to this headline">&nbsp;&pi;</a>'.format(uuid.uuid4()))
```

Comment:
You’re doing UUIDs too much. I would:
* check if the header has an `id` already, and use that
* otherwise, create a slugified form of the header and use that, watching out for conflicts within the same page and appending numbers [(example for Django)](https://github.com/Kwpolska/achieve/blob/master/achieve/helpers.py#L47)